bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/internal/events/EventHandler.java

```diff
@@ -0,0 +1,148 @@
+import org.eclipse.smarthome.core.events.Event;
+import org.eclipse.smarthome.core.events.EventFactory;
+import org.eclipse.smarthome.core.events.EventFilter;
+import org.eclipse.smarthome.core.events.EventSubscriber;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Handle Eclipse SmartHome events encapsulated by OSGi events.
+ *
+ * @author Markus Rathgeb - Initial contribution
+ */
+@NonNullByDefault
+public class EventHandler {
+
+    private final Logger logger = LoggerFactory.getLogger(EventHandler.class);
+
+    private final ConcurrentHashMap<String, CopyOnWriteArraySet<EventSubscriber>> typedEventSubscribers;
+    private final Map<String, EventFactory> typedEventFactories;
+    private final SafeCaller safeCaller;
+
+    /**
+     * Create a new event handler.
+     *
+     * @param typedEventSubscribers the event subscribers indexed by the event type
+     * @param typedEventFactories the event factories indexed by the event type
+     * @param safeCaller the safe caller to use
+     */
+    public EventHandler(final ConcurrentHashMap<String, CopyOnWriteArraySet<EventSubscriber>> typedEventSubscribers,
+            final Map<String, EventFactory> typedEventFactories, final SafeCaller safeCaller) {
+        this.typedEventSubscribers = typedEventSubscribers;
+        this.typedEventFactories = typedEventFactories;
+        this.safeCaller = safeCaller;
+    }
+
+    public void handleEvent(org.osgi.service.event.Event osgiEvent) {
+        logger.trace("Handle OSGi event (event: {})", osgiEvent);
+
+        Object typeObj = osgiEvent.getProperty("type");
+        Object payloadObj = osgiEvent.getProperty("payload");
+        Object topicObj = osgiEvent.getProperty("topic");
+        Object sourceObj = osgiEvent.getProperty("source");
+
+        if (typeObj instanceof String && payloadObj instanceof String && topicObj instanceof String) {
+            String typeStr = (String) typeObj;
+            String payloadStr = (String) payloadObj;
+            String topicStr = (String) topicObj;
+            String sourceStr = (sourceObj instanceof String) ? (String) sourceObj : null;
+            if (!typeStr.isEmpty() && !payloadStr.isEmpty() && !topicStr.isEmpty()) {
+                handleEvent(typeStr, payloadStr, topicStr, sourceStr);
+            }
+        } else {
+            logger.error(
+                    "The handled OSGi event is invalid. Expect properties as string named 'type', 'payload' and 'topic'. "
+                            + "Received event properties are: {}",
+                    Arrays.toString(osgiEvent.getPropertyNames()));
+        }
+    }
+
+    private void handleEvent(final String type, final String payload, final String topic,
+            final @Nullable String source) {
+        final EventFactory eventFactory = typedEventFactories.get(type);
+        if (eventFactory == null) {
+            logger.warn("Could not find an Event Factory for the event type '{}'.", type);
```

Comment:
Could we reduce this to debug again (as in the current master)?

Reply:
If you want :wink: